document-store/src/main/java/org/hypertrace/core/documentstore/expression/impl/DocumentArrayFilterExpression.java

```diff
@@ -0,0 +1,68 @@
+package org.hypertrace.core.documentstore.expression.impl;
+
+import com.google.common.base.Preconditions;
+import lombok.AccessLevel;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Value;
+import org.hypertrace.core.documentstore.expression.operators.ArrayOperator;
+import org.hypertrace.core.documentstore.expression.type.FilterTypeExpression;
+import org.hypertrace.core.documentstore.expression.type.SelectTypeExpression;
+import org.hypertrace.core.documentstore.parser.FilterTypeExpressionVisitor;
+
+/**
+ * Expression representing a condition for filtering on document array fields
+ *
+ * <p>Example: If product is an array field (containing documents)<code>
+ * ANY(product) [color IN ('Blue', 'Green') AND color != 'Black' AND name = 'Comb']
+ * </code> can be constructed as <code>
+ *   DocumentArrayFilterExpression.builder()
+ *    .operator(ANY)
+ *    .arraySource(IdentifierExpression.of("product"))
+ *    .filter(
+ *      LogicalExpression.and(
+ *        RelationalExpression.of(
+ *          IdentifierExpression.of("color"),
+ *          IN,
+ *          ConstantExpression.ofStrings("Blue", "Green")),
+ *        RelationalExpression.of(
+ *          IdentifierExpression.of("color"),
+ *          NEQ,
+ *          ConstantExpression.of("Black")),
+ *        RelationalExpression.of(
+ *          IdentifierExpression.of("name"),
+ *          EQ,
+ *          ConstantExpression.of("Comb"))
+ *      )
+ *    .build();
+ */
+@Value
+@Builder
+@AllArgsConstructor(access = AccessLevel.PRIVATE)
+public class DocumentArrayFilterExpression implements FilterTypeExpression {
```

Comment:
Other filter expressions are not prefixed with `Document`, any specific reason?

Could this be just called - `ArrayExpression` or `ArrayFilterExpression`?

Reply:
> any specific reason?

Yeah. To clearly distinguish between a primitive-array and a document-array (object-array).
Calling it a document array instead of an object array, because it supports filtering by nested fields too.